aspnetcore/tutorials/build-your-first-razor-components-app.md

```diff
@@ -0,0 +1,188 @@
+
+   *WebApplication1.App/Pages/Counter.cshtml*:
+
+   [!code-cshtml[](build-your-first-razor-components-app/samples/3.x/WebApplication1/WebApplication1.App/Pages/Counter.cshtml?highlight=12,16)]
+
+   > [!NOTE]
+   > From Visual Studio, you can quickly add a component parameter by using the `para` snippet. Type `para` and press the `Tab` key twice.
+
+1. Specify an `IncrementAmount` parameter in the Home component's `<Counter>` element using an attribute. Set the value to increment the counter by ten.
+
+   *WebApplication1.App/Pages/Index.cshtml*:
+
+   [!code-cshtml[](build-your-first-razor-components-app/samples/3.x/WebApplication1/WebApplication1.App/Pages/Index.cshtml?highlight=7)]
+
+1. Reload the page. The home page counter increments by ten each time the **Click me** button is selected. The counter on the *Counter* page increments by one.
+
+## Route to components
+
+The `@page` directive at the top of the *Counter.cshtml* file specifies that this component is a routing endpoint. The `Counter` component handles requests sent to `/Counter`. Without the `@page` directive, the component doesn't handle routed requests, but the component can still be used by other components.
+
+## Dependency injection
+
+Services registered in the app's service container are available to components via [dependency injection (DI)](xref:fundamentals/dependency-injection). Inject services into a component using the `@inject` directive.
+
+Examine the directives of the `FetchData` component (*WebApplication1.App/Pages/FetchData.cshtml*). The `@inject` directive is used to inject the instance of the `WeatherForecastService` service into the component:
+
+[!code-cshtml[](build-your-first-razor-components-app/samples_snapshot/3.x/FetchData1.cshtml?highlight=3)]
+
+The `WeatherForecastService` service is registered as a [singleton](xref:fundamentals/dependency-injection#service-lifetimes), so one instance of the service is available throughout the app.
+
+The `FetchData` component uses the injected service, as `ForecastService`, to retrieve an array of `WeatherForecast` objects:
+
+[!code-cshtml[](build-your-first-razor-components-app/samples_snapshot/3.x/FetchData2.cshtml?highlight=6)]
+
+An [@foreach](/dotnet/csharp/language-reference/keywords/foreach-in) loop is used to render each forecast instance as a row in the table of weather data:
+
+[!code-cshtml[](build-your-first-razor-components-app/samples_snapshot/3.x/FetchData3.cshtml?highlight=11-19)]
+
+## Build a todo list
+
+Add a new page to the app that implements a simple todo list.
+
+1. Add an empty file to the *WebApplication1.App/Pages* folder named *Todo.cshtml*.
+
+1. Provide the initial markup for the page:
+
+   ```cshtml
+   @page "/todo"
+
+   <h1>Todo</h1>
+   ```
+
+1. Add the Todo page to the navigation bar by updating *WebApplication1.App/Shared/NavMenu.cshtml*. Add a `<NavLink>` for the Todo page by adding the following list item markup below the existing list items.
```

Comment:
Consider adding  some text here explaining that the `NavMenu` component is used in the app's layout, and describe a bit how Razor Component layouts work.

Reply:
Or maybe add a section above that describes how layouts work in the app.